File: BinaryNotes.NET/BinaryNotesTests/org/bn/performance/DummyPerformanceTest.cs
/*
 Copyright 2006-2011 Abdulla Abdurakhmanov (abdulla@latestbit.com)
 Original sources are available at www.latestbit.com

 Licensed under the Apache License, Version 2.0 (the "License");
 you may not use this file except in compliance with the License.
 You may obtain a copy of the License at

 http://www.apache.org/licenses/LICENSE-2.0

 Unless required by applicable law or agreed to in writing, software
 distributed under the License is distributed on an "AS IS" BASIS,
 WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 See the License for the specific language governing permissions and
 limitations under the License.
 */

using Microsoft.VisualStudio.TestTools.UnitTesting;
using org.bn.coders;
using org.bn.coders.ber;
using org.bn.coders.per;
using org.bn.coders.test_asn;
using System;

namespace org.bn.performance
{
    [TestClass]
    public class DummyPerformanceTest
    {
        protected void runEncoderPerfTest(string encoding)
        {
            IEncoder encoder = CoderFactory.getInstance().newEncoder(encoding);
            Assert.IsNotNull(encoder);
            // Create test structure
            DataSeq dt = new BERCoderTestUtils().createDataSeq();
            System.IO.Stream stream = new System.IO.MemoryStream();
            // Start test
            DateTime startTime = System.DateTime.Now;
            for (int i = 0; i < 100; i++)
            {
                encoder.encode<DataSeq>(dt, stream);
            }
            DateTime endTime = System.DateTime.Now;
            TimeSpan interval = (endTime-startTime);
            System.Console.WriteLine("Encode elapsed time for " + encoding + ": " + interval.TotalSeconds );
        }

        protected void runDecoderPerfTest(string encoding, CoderTestUtilities coderUtils)
        {
            IDecoder encoder = CoderFactory.getInstance().newDecoder(encoding);
            Assert.IsNotNull(encoder);
            // Create test structure
            System.IO.Stream stream = new System.IO.MemoryStream(
                    coderUtils.createDataSeqBytes()
            );
            // Start test
            DateTime startTime = System.DateTime.Now;
            for (int i = 0; i < 100; i++)
            {
                DataSeq dt = encoder.decode<DataSeq>(stream);
                stream.Position = 0;
            }
            DateTime endTime = System.DateTime.Now;
            TimeSpan interval = (endTime - startTime);
            System.Console.WriteLine("Decode elapsed time for " + encoding + ": " + interval.TotalSeconds);
        }

        [TestMethod]
        public void testEncodePerf()
        {
            runEncoderPerfTest("BER");
            runEncoderPerfTest("PER");
            runEncoderPerfTest("PER/Unaligned");
        }

        [TestMethod]
        public void testDecodePerf(){
            runDecoderPerfTest("BER", new BERCoderTestUtils());
            runDecoderPerfTest("PER", new PERAlignedCoderTestUtils());
            runDecoderPerfTest("PER/Unaligned", new PERUnalignedCoderTestUtils());
        }
    }
}

File: JavaLibrary/src/test/java/org/bn/performance/DummyPerformanceTest.java
/*
 Copyright 2006-2011 Abdulla Abdurakhmanov (abdulla@latestbit.com)
 
 Licensed under the Apache License, Version 2.0 (the "License");
 you may not use this file except in compliance with the License.
 You may obtain a copy of the License at

 http://www.apache.org/licenses/LICENSE-2.0

 Unless required by applicable law or agreed to in writing, software
 distributed under the License is distributed on an "AS IS" BASIS,
 WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 See the License for the specific language governing permissions and
 limitations under the License.
 */
package org.bn.performance;

import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;
import org.bn.CoderFactory;
import org.bn.IDecoder;
import org.bn.IEncoder;
import org.bn.coders.CoderTestUtilities;
import org.bn.coders.ber.BERCoderTestUtils;
import org.bn.coders.per.PERAlignedCoderTestUtils;
import org.bn.coders.per.PERUnalignedCoderTestUtils;
import org.bn.coders.test_asn.*;
import static org.junit.Assert.assertNotNull;
import org.junit.Test;

public class DummyPerformanceTest {
    
    protected void runEncoderPerfTest(String encoding) throws Exception {
        IEncoder encoder = CoderFactory.getInstance().newEncoder(encoding);
        assertNotNull(encoder);
        // Create test structure
        DataSeq dt = new BERCoderTestUtils().createDataSeq();
        ByteArrayOutputStream stream = new ByteArrayOutputStream();
        // Start test
        long startTime = System.currentTimeMillis();
        for (int i = 0; i < 100; i++)
        {
            encoder.encode(dt, stream);
        }
        long endTime = System.currentTimeMillis();
        long interval = (endTime-startTime);
        System.out.println("Encode elapsed time for " + encoding + ": " + interval/1000.0 );
    }

    protected void runDecoderPerfTest(String encoding, CoderTestUtilities coderUtils) throws Exception {
        IDecoder decoder = CoderFactory.getInstance().newDecoder(encoding);
        assertNotNull(decoder);
        // Create test structure
        ByteArrayInputStream stream = new ByteArrayInputStream( coderUtils.createDataSeqBytes() );
        // Start test
        long startTime = System.currentTimeMillis();
        for (int i = 0; i < 100; i++)
        {
            DataSeq dt = decoder.decode(stream, DataSeq.class);
            stream.reset();
        }
        long endTime = System.currentTimeMillis();
        long interval = (endTime-startTime);
        System.out.println("Decode elapsed time for " + encoding + ": " + interval/1000.0 );
    }
    
    @Test
    public void testEncodePerf() throws Exception {
        runEncoderPerfTest("BER");
        runEncoderPerfTest("PER");
        runEncoderPerfTest("PER/Unaligned");
    }

    @Test
    public void testDecodePerf() throws Exception {
        runDecoderPerfTest("BER", new BERCoderTestUtils());
        runDecoderPerfTest("PER", new PERAlignedCoderTestUtils());
        runDecoderPerfTest("PER/Unaligned", new PERUnalignedCoderTestUtils());
    }
}